### ISSUE TITLE
ENOENT privkey.pem

### ISSUE THREAD

**Issue by tremby** [Daplie/letsencrypt-express#45]
I'm trying to give this a go.

I'm in testing mode, and I'm getting an `ENOENT` error. It's looking for `letsencrypt-path/live/my-domain/privkey.pem`, and it's right that it doesn't exist. This error comes up a couple more times, and the process then goes on to fail.

Was I supposed to set something up in this `letsencrypt-path` directory beforehand?

**Comment by coolaj86** [comment]
does `letsencrypt-path` exist? are there any permission errors (you're running as one user, but the parent folder is owned by another user)?

**Comment by tremby** [comment]
letsencrypt-express is making and populating it, so I can't see that it could be a permissions thing.

```
$ tree letsencrypt-path
letsencrypt-path
├── accounts
│   └── acme-staging.api.letsencrypt.org
│       └── directory
│           └── 572fd36ef6bb5167412f6c76cbf9cf01
│               ├── meta.json
│               ├── private_key.json
│               └── regr.json
└── renewal
    ├── my-domain.conf
    └── my-domain.conf.bak

5 directories, 5 files
```

**Comment by Bluejay47** [comment]
Same issue here on Windows 10 64bit if that matters...

**Comment by RobertHerhold** [comment]
Ran into this as well

**Comment by RobertHerhold** [comment]
@coolaj86 Any resolution for this?

**Comment by tremby** [comment]
I haven't found a solution yet, though I haven't been looking very hard.

**Comment by RadianSmile** [comment]
Hope this issue could be fixed ASAP because this package is just totally broken with this issue.

**Comment by coolaj86** [comment]
I haven't been able to reproduce this as of yet.

Can you guess chat a little bit and figure out what your common thread is? Maybe operating system or node version or platform or something like that?

**Comment by twktue** [comment]
I am running into this problem using Node v6.3.1 on Ubuntu 14.04. I created a blank application just to test this working and I can't seem to get it to work. What versions of Node are tested and working?

I'm attaching both the script I am trying to run and the output its creating after I attempt to hit the server using HTTPS.

[app.js.txt](https://github.com/Daplie/letsencrypt-express/files/404755/app.js.txt)
[ENOENT-output.txt](https://github.com/Daplie/letsencrypt-express/files/404756/ENOENT-output.txt)

I hope that helps. I really wanted this to work and be a 5 minute experience...

**Comment by polyma** [comment]
Having the same issue on Ubuntu 14.04 and Node v6.3.1.

Permissions of ~/letsencrypt/etc set to relevant user (recursively).

It seems to work in non-testing mode but produces an invalid certificate (NET::ERR_CERT_COMMON_NAME_INVALID) as the certificate is run from localhost.daplie.com

**Comment by polyma** [comment]
The issue seems to be with this commit of letsencrypt/lib/core.js https://github.com/Daplie/node-letsencrypt/commit/4a389606b058ce1bc447550f42033883bb38fc30#diff-f265d6a6b6b64d4d096fddcf7d620a76

  ```
return fetchFromConfigLiveDir(args).then(function (certs) {
    var halfLife = (certs.expiresAt - certs.issuedAt) / 2;

    if (!certs || (Date.now() - certs.issuedAt) > halfLife) {
      // There is no cert available
      // Or the cert is more than half-expired
      return getCertificateAsync(args, defaults, handlers);
    }
```

the halfLife variable accesses certs that that may not be present. I fixed this by moving the halflife declaration into the if statement:

  ```
return fetchFromConfigLiveDir(args).then(function (certs) {
    if (!certs || (Date.now() - certs.issuedAt) > ((certs.expiresAt - certs.issuedAt) / 2)) {
      // There is no cert available
      // Or the cert is more than half-expired
      return getCertificateAsync(args, defaults, handlers);
    }
```

**Comment by coolaj86** [comment]
@polyma that's actually unrelated to this issue, but thank you for finding that. Your issue is fixed as of v1.5.4.

**Comment by coolaj86** [comment]
@twktue your issue was the same as @polyma's and has been fixed.

**Comment by coolaj86** [comment]
Everybody else: You're all on Windows, right?

I think I found the problem. The default paths are all unix pathnames and you probably didn't set them to use Windows pathnames - hence only windows people using all of the default settings are having issues.

In the new version I'm about to publish all pathnames are created os-specific, so you shouldn't have this issue anymore.

I'll update again when I publish.

**Comment by tremby** [comment]
I'm on Linux. The issue may or may not be fixed for me; I'm afraid I don't have time to test at present. After my current set of deadlines hopefully I'll circle back to this.

**Comment by coolaj86** [comment]
Just published letsencrypt-express v2.0.0, based on node-letsencrypt v2.1.0 and the mixed paths issue that was causing ENOENT has been completely solved.

If you are still having this issue please open a new bug, as it is related to something else.

See the new README.md
